sdk/src/marginCalculation.ts

```diff
@@ -0,0 +1,351 @@
+	}
+
+	setFuelSpotDelta(marketIndex: number, delta: BN): this {
+		this.fuelSpotDeltas = [{ marketIndex, delta }];
+		return this;
+	}
+
+	setFuelSpotDeltas(deltas: Array<{ marketIndex: number; delta: BN }>): this {
+		this.fuelSpotDeltas = deltas;
+		return this;
+	}
+
+	setFuelNumerator(numerator: BN): this {
+		this.fuelBonusNumerator = numerator ?? new BN(0);
+		return this;
+	}
+
+	setMarginRatioOverride(ratio: number): this {
+		this.marginRatioOverride = ratio;
+		return this;
+	}
+
+	trackMarketMarginRequirement(marketIdentifier: MarketIdentifier): this {
+		if (this.mode.type !== 'Liquidation') {
+			throw new Error(
+				'InvalidMarginCalculation: Cant track market outside of liquidation mode'
+			);
+		}
+		this.mode.marketToTrackMarginRequirement = marketIdentifier;
+		return this;
+	}
+}
+
+export class IsolatedMarginCalculation {
+	marginRequirement: BN;
+	totalCollateral: BN; // deposit + pnl
+	totalCollateralBuffer: BN;
+	marginRequirementPlusBuffer: BN;
+
+	constructor() {
+		this.marginRequirement = new BN(0);
+		this.totalCollateral = new BN(0);
+		this.totalCollateralBuffer = new BN(0);
+		this.marginRequirementPlusBuffer = new BN(0);
+	}
+
+	getTotalCollateralPlusBuffer(): BN {
+		return this.totalCollateral.add(this.totalCollateralBuffer);
+	}
+
+	meetsMarginRequirement(): boolean {
+		return this.totalCollateral.gte(this.marginRequirement);
+	}
+
+	meetsMarginRequirementWithBuffer(): boolean {
+		return this.getTotalCollateralPlusBuffer().gte(
+			this.marginRequirementPlusBuffer
+		);
+	}
+
+	marginShortage(): BN {
+		const shortage = this.marginRequirementPlusBuffer.sub(
+			this.getTotalCollateralPlusBuffer()
+		);
+		return shortage.isNeg() ? new BN(0) : shortage;
+	}
+}
+
+export class MarginCalculation {
+	context: MarginContext;
+	totalCollateral: BN;
+	totalCollateralBuffer: BN;
+	marginRequirement: BN;
+	marginRequirementPlusBuffer: BN;
+	isolatedMarginCalculations: Map<number, IsolatedMarginCalculation>;
+	numSpotLiabilities: number;
+	numPerpLiabilities: number;
+	allDepositOraclesValid: boolean;
+	allLiabilityOraclesValid: boolean;
+	withPerpIsolatedLiability: boolean;
+	withSpotIsolatedLiability: boolean;
+	trackedMarketMarginRequirement: BN;
+	fuelDeposits: number;
+	fuelBorrows: number;
+	fuelPositions: number;
+
+	constructor(context: MarginContext) {
+		this.context = context;
+		this.totalCollateral = new BN(0);
+		this.totalCollateralBuffer = new BN(0);
+		this.marginRequirement = new BN(0);
+		this.marginRequirementPlusBuffer = new BN(0);
+		this.isolatedMarginCalculations = new Map();
+		this.numSpotLiabilities = 0;
+		this.numPerpLiabilities = 0;
+		this.allDepositOraclesValid = true;
+		this.allLiabilityOraclesValid = true;
+		this.withPerpIsolatedLiability = false;
+		this.withSpotIsolatedLiability = false;
+		this.trackedMarketMarginRequirement = new BN(0);
+		this.fuelDeposits = 0;
+		this.fuelBorrows = 0;
+		this.fuelPositions = 0;
+	}
+
+	addIsolatedTotalCollateral(delta: BN): void {
```

Comment:
this was a bug in program, should be addCrossMarginTotalCollateral and addCrossMarginRequirement